### PR TITLE
[enhance](orc) Optimize ORC Predicate Pushdown for OR-connected Predicate

### PR DESCRIPTION
**Pull request by suxiaogang223**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:
This issue addresses a limitation in Apache Doris where only predicates joined by AND are pushed down to the ORC reader, leaving OR-connected predicates unoptimized. By extending pushdown functionality to handle these OR conditions, the aim is to better leverage ORC’s predicate pushdown capabilities, reducing data reads and improving query performance.

### Check List (For Committer)

- Test 

    - [x] Regression test
    - [x] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No colde files have been changed.
        - [ ] Other reason 

- Behavior changed:

    - [x] No.
    - [ ] Yes. 

- Does this need documentation?

    - [x] No.
    - [ ] Yes. 

- Release note

    
    
    None

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by suxiaogang223** [comment]
run buildall

**Comment by suxiaogang223** [comment]
run buildall

**Comment by suxiaogang223** [comment]
run buildall

**Comment by suxiaogang223** [comment]
run buildall

**Comment by suxiaogang223** [comment]
run external

**Comment by suxiaogang223** [comment]
run external

**Comment by suxiaogang223** [comment]
run external

**Comment by suxiaogang223** [comment]
run buildall

File: be/src/vec/exec/format/orc/vorc_reader.cpp [hunk 1]
```diff
@@ -558,190 +562,316 @@ std::tuple<bool, orc::Literal> convert_to_orc_literal(const orc::Type* type, con
...
-
-    if (unsupported_pushdown_types != nullptr) {
-        for (vector<orc::TypeKind>::iterator it = unsupported_pushdown_types->begin();
-             it != unsupported_pushdown_types->end(); ++it) {
-            if (*it == type->getKind()) {
-                // Unsupported type
-                return predicates;
-            }
-        }
+std::tuple<bool, orc::Literal, orc::PredicateDataType> OrcReader::_make_orc_leteral(
```
Comment (morningman) [inline be/src/vec/exec/format/orc/vorc_reader.cpp:565]:
```suggestion
std::tuple<bool, orc::Literal, orc::PredicateDataType> OrcReader::_make_orc_literal(
```

File: be/src/vec/exec/format/orc/vorc_reader.cpp [hunk 2]
```diff
@@ -558,190 +562,316 @@ std::tuple<bool, orc::Literal> convert_to_orc_literal(const orc::Type* type, con
...
-    for (auto& col_name : _lazy_read_ctx.all_read_columns) {
-        auto iter = colname_to_value_range->find(col_name);
-        if (iter == colname_to_value_range->end()) {
-            continue;
-        }
-        auto type_it = type_map.find(_col_name_to_file_col_name[col_name]);
-        if (type_it == type_map.end()) {
-            continue;
+
+    auto sargBuilder = orc::SearchArgumentFactory::newBuilder();
```
Comment (morningman) [inline be/src/vec/exec/format/orc/vorc_reader.cpp:860]:
```suggestion
    auto sarg_builder = orc::SearchArgumentFactory::newBuilder();
```

File: be/src/vec/exec/format/orc/vorc_reader.cpp [hunk 3]
```diff
@@ -558,190 +562,316 @@ std::tuple<bool, orc::Literal> convert_to_orc_literal(const orc::Type* type, con
...
-            }
-        }
-        if (!in_predicate.literals.empty()) {
-            predicates.emplace_back(in_predicate);
+// check if the expr can be pushed down to orc reader
+bool OrcReader::_check_expr_can_push_down(const VExprSPtr& expr) {
+    DCHECK_NOTNULL(expr);
+    switch (expr->op()) {
+    case TExprOpcode::COMPOUND_AND:
+    case TExprOpcode::COMPOUND_OR:
```
Comment (morningman) [inline be/src/vec/exec/format/orc/vorc_reader.cpp:615]: I have a question that, for example: `A or B`. if `B` can not be pushed down, then `A or B` can not be pushed down.
Why you say that for `OR`, if "at least one child can be pushed down", then return true?

File: be/src/vec/exec/format/orc/vorc_reader.cpp [hunk 4]
```diff
@@ -558,190 +562,316 @@ std::tuple<bool, orc::Literal> convert_to_orc_literal(const orc::Type* type, con
...
-            continue;
-        }
-        auto type_it = type_map.find(_col_name_to_file_col_name[col_name]);
-        if (type_it == type_map.end()) {
-            continue;
+
+    auto sargBuilder = orc::SearchArgumentFactory::newBuilder();
+    sargBuilder->startAnd();
+    for (const auto& expr_ctx : conjuncts) {
+        if (!_build_search_argument(expr_ctx->root(), sargBuilder)) {
```
Comment (morningman) [inline be/src/vec/exec/format/orc/vorc_reader.cpp:863]: if there is A、B、C in `conjuncts`, and if `_build_search_argument(A)` return false, then we no longer try to push down B and C?

File: be/src/vec/exec/format/orc/vorc_reader.cpp [hunk 5]
```diff
@@ -558,190 +562,316 @@ std::tuple<bool, orc::Literal> convert_to_orc_literal(const orc::Type* type, con
...
+        DCHECK_EQ(expr->children().size(), 1);
+        if (!_build_search_argument(expr->children()[0], builder)) {
+            return false;
+        }
+        builder->end();
+        break;
+    case TExprOpcode::GE: {
+        builder->startNot();
+        DCHECK(expr->children().size() == 2);
+        DCHECK(expr->children()[0]->is_slot_ref());
```
Comment (morningman) [inline be/src/vec/exec/format/orc/vorc_reader.cpp:691]: How to make sure that first child must be slot and second must be literal?

File: be/src/vec/exec/format/orc/vorc_reader.cpp [hunk 6]
```diff
@@ -558,190 +562,316 @@ std::tuple<bool, orc::Literal> convert_to_orc_literal(const orc::Type* type, con
...
+        __builtin_unreachable();
+    }
     }
     return true;
 }
 
-bool OrcReader::_init_search_argument(
-        std::unordered_map<std::string, ColumnValueRangeType>* colname_to_value_range) {
-    if ((!_enable_filter_by_min_max) || colname_to_value_range->empty()) {
+bool OrcReader::_init_search_argument(const VExprContextSPtrs& conjuncts) {
```
Comment (morningman) [inline be/src/vec/exec/format/orc/vorc_reader.cpp:848]: For a certain query, the conjuncts is identical in all orc readers.
So I think we can just init search argument once and use the result for every orc reader?

Comment (suxiaogang223) [reply be/src/vec/exec/format/orc/vorc_reader.cpp:848]: it is possible to cache searchArgument into member variables of OrcReader

File: be/src/vec/exec/format/orc/vorc_reader.cpp (hunk 3 above)
Comment (suxiaogang223) [reply be/src/vec/exec/format/orc/vorc_reader.cpp:615]: Indeed a problem, need to check that all children can be pushed down for "or".

File: be/src/vec/exec/format/orc/vorc_reader.cpp (hunk 4 above)
Comment (suxiaogang223) [reply be/src/vec/exec/format/orc/vorc_reader.cpp:863]: Indeed a problem, return false only if all conjuncts  cannot be _build_search_argument normally.

**Comment by suxiaogang223** [comment]
run external

**Comment by suxiaogang223** [comment]
run buildall

**Comment by suxiaogang223** [comment]
run buildall

**Comment by suxiaogang223** [comment]
run buildall

File: be/src/vec/exec/format/orc/vorc_reader.cpp [hunk 7]
```diff
@@ -558,190 +565,347 @@ std::tuple<bool, orc::Literal> convert_to_orc_literal(const orc::Type* type, con
...
-        if (predicate.literals.size() == 1) {
-            builder->equals(predicate.col_name, predicate.data_type, predicate.literals[0]);
-        } else {
-            builder->in(predicate.col_name, predicate.data_type, predicate.literals);
+
+    switch (expr->op()) {
+    case TExprOpcode::COMPOUND_AND:
+        builder->startAnd();
+        for (const auto& child : expr->children()) {
+            if (!_build_search_argument(child, builder)) {
```
Comment (morningman) [inline be/src/vec/exec/format/orc/vorc_reader.cpp:788]: I think for AND, it should be `continue`?

**Comment by suxiaogang223** [comment]
run buildall

**Comment by suxiaogang223** [comment]
run buildall

**Comment by suxiaogang223** [comment]
run buildall

**Comment by suxiaogang223** [comment]
run external

**Review by morningman — APPROVED** [review]
LGTM

**Review by kaka11chen — APPROVED** [review]
LGTM

**Pull request merged**